lib/carbon/cache.py

```diff
@@ -106,29 +121,40 @@ def _generate_queue():
         if settings.LOG_CACHE_QUEUE_SORTS and size:
           log.msg("Sorted %d cache queues in %.6f seconds" % (size, time.time() - t))
         while metric_counts:
-          yield itemgetter(0)(metric_counts.pop())
+          (metric_name, numPoints) = metric_counts.pop()
+          if self.mdpu_flag:
+            if numPoints >= settings.MIN_DATAPOINTS_PER_UPDATE:
+              yield metric_name
+            else:
+              metric_counts = []
+              yield None
+          else:
+            yield metric_name
         if settings.LOG_CACHE_QUEUE_SORTS and size:
           log.msg("Queue consumed in %.6f seconds" % (time.time() - t))
 
     self.queue = _generate_queue()
 
-  def choose_item(self):
+  def choose_item(self, mdpu=False):
+    self.mdpu_flag = mdpu
     return next(self.queue)
 
 
 class TimeSortedStrategy(DrainStrategy):
-  """ This strategy prefers metrics wich are lagging behind
+  """ This strategy prefers metrics which are lagging behind
   guarantees every point gets written exactly once during
   a loop of the cache """
   def __init__(self, cache):
     super(TimeSortedStrategy, self).__init__(cache)
+    self.mdpu_flag = False
 
     def _generate_queue():
       while True:
         t = time.time()
         metric_lw = sorted(self.cache.watermarks, key=lambda x: x[1], reverse=True)
         if settings.MIN_TIMESTAMP_LAG:
-          metric_lw = [x for x in metric_lw if t - x[1] > settings.MIN_TIMESTAMP_LAG]
+          metric_lw = [x for x in metric_lw if t - x[1] > settings.MIN_TIMESTAMP_LAG
+                        or (self.mdpu_flag and x[1] >= settings.MIN_DATAPOINTS_PER_UPDATE)]
```

Comment:
Ah, that's should be fixed too - with default MIN_DATAPOINTS_PER_UPDATE == 1 this effectively turns timesorted strategy to sorted.